crates/bevy_app/src/preferences.rs

```diff
@@ -0,0 +1,275 @@
+use bevy_ecs::system::Resource;
+use bevy_reflect::{Map, Reflect, TypePath};
+
+use crate::Plugin;
+
+/// Adds application [`Preferences`] functionality.
+pub struct PreferencesPlugin;
+
+impl Plugin for PreferencesPlugin {
```

Comment:
Why a centralized Preferences store instead of using the ECS?

```rust
#[derive(Reflect)]
struct MySettings {
  foo: Foo,
  bar: Bar,
}

app.add_plugin(SettingsPlugin::<MySettings>()::with_path("path/to/my_settings.ron"))

// later
fn system(my_settings: Res<MySettings>) {
}
```

Can we / should we use the asset system for this?

```rust
#[derive(Asset, Reflect)]
struct MySettings {
  foo: Foo,
  bar: Bar,
} 

// this kicks off an asset load
app.add_plugin(SettingsPlugin::<MySettings>()::with_path("path/to/my_settings.ron"))
```

Reply:
That's interesting, but neither your proposal or this PR really solves the problem I was hoping to solve. What I was hoping for is something that would abstract away platform differences - browser local storage vs `${HOME}/.config` vs `${HOME}/Libraries/Application Support` vs `C:\\Program Files\etc.`.

* For some platforms, the game installation dir might not be writable (I think?).
* For some platforms (browser), there might not even be a filesystem.

My specific use case is that I have a color picker that wants to save a list of "recent colors" as a user preference. Because this is part of a library, I have no knowledge of what platforms it will eventually run on, that's up to the game author that uses my crate. So I need some way to read/write my preferences, without knowing any of the details of how they will be stored.

Reply:
Using the asset system could play into solving this problem, via a new `config` AssetSource, that writes to whatever config location makes sense for a given platform. I've proposed this exact API very recently on discord:

```rust
// ex: loads from user `AppData` on windows, `.config` on linux, etc
asset_server.load("config://my_settings.ron")
```

Reply:
(you could build this today as a 3rd party plugin in Bevy Asset V2)

Reply:
Very possibly, yes. I think this would be fruitful to add as part of the discussion in your other comment. I definitely heard my gut saying "use assets", but I responded "idk how the heck to use them".

In my mind, this could be provided by the user or plugins, but built on top of this centralized map. Want to save preferences to json? Use `bevy_json_preferences`, which might have other features like splitting into multiple files, or filtering before serialization.

To me this is orthogonal to providing a preferences API. The problem this PR solves is "how do we get the ecosystem to agree on a central store for globally persistent state with a dynamic format". What to use for serialization seems like a whole new can of worms.

Reply:
>  What to use for serialization seems like a whole new can of worms.

I agree that it can (and should) be a separate discussion.

> To me this is orthogonal to providing a preferences API. The problem this PR solves is "how do we get the ecosystem to agree on a central store for globally persistent state with a dynamic format".

To this point, can you explain why a centralized `Preferences` store is necessary at all relative to storing each preference in individual Resources (as I discussed above)? Using resources seems preferential from a "consume from code" perspective.

Reply:
You could implement this API in a way where the in-memory state lives in Resources; I don't think the two are mutually exclusive.

The discussion that spawned this came from a desire to have some blessed solution that allows us to read/write a single user-preferences file that can be made persistent. I don't _really_ care where that data lives in memory, the only bit of this design that actually matters to the user is that it provides a singular place for plugin devs to declare "I want this data to be globally persistent and user configurable". 

Using resources is an implementation detail - if we want to turn this into something that needs "gather" and "distribute" phases, that's fine, but it will be more code and implementation complexity. The end goal is a single file, so spreading the in-memory representation across the ECS in multiple Resources is possible, but annoying.

So, to answer your question in the most long-winded way possible, the backing storage for `Preferences` could go into `Resource`s for all I care, but I do think there is value in having a centralized API surface for plugins to declare that they have persistent, user/dev-configurable settings. It might be that this is not the right place to put that API! In this comment, I suggest it might make more sense for it to be a trait extension on `Plugin`.